machine-charm/src/charm.py

```diff
@@ -52,6 +49,12 @@ def _on_controller_relation_changed(self, event):
 
         self._configure(**requires.data)
 
+    def _on_config_changed(self, _):
+        relation = self.model.get_relation("controller")
+        if relation:
+            data = JujuDashData(relation.data[relation.app])
+            self._configure(data['controller_url'], data['identity_provider_url'], data['is_juju'])
```

Comment:
In your specific use case, you wanted to be able to rely on the controller_url in the config, rather than the controller_url in the relation, correct? 

To handle that use case, you probably want to do a no-op if the relation exists, because the relation_changed handler will set your config, and inspect the config values if the relation *doesn't* exist ...

Reply:
if i set the dns-name config option, that will get picked up in the _configure method on L86.. so i still want to use controller_url and other parameters from the controller relation, but i also want to run _configure on any config change to pick up the possibly changed dns-name..

Reply:
Got it. That makes sense!